src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -299,33 +334,208 @@ private BoundExpression RewriteStringConcatenationOneExpr(BoundExpression lowere
+
+            if (secondIsChar)
+            {
+                loweredSecond = WrapCharExprIntoToStringCall(loweredSecond);
+            }
+
+            if (thirdIsChar)
+            {
+                loweredThird = WrapCharExprIntoToStringCall(loweredThird);
+            }
 
             var method = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_String__ConcatStringStringString);
             Debug.Assert((object)method != null);
 
             return BoundCall.Synthesized(syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, method, ImmutableArray.Create(loweredFirst, loweredSecond, loweredThird));
         }
 
-        private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, BoundExpression loweredFirst, BoundExpression loweredSecond, BoundExpression loweredThird, BoundExpression loweredFourth)
+        private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, BoundExpression loweredFirst, BoundExpression loweredSecond, BoundExpression loweredThird, BoundExpression loweredFourth, ArrayBuilder<LocalSymbol> previousLocals)
         {
-            Debug.Assert(loweredFirst.HasAnyErrors || loweredFirst.Type is { } && loweredFirst.Type.IsStringType());
-            Debug.Assert(loweredSecond.HasAnyErrors || loweredSecond.Type is { } && loweredSecond.Type.IsStringType());
-            Debug.Assert(loweredThird.HasAnyErrors || loweredThird.Type is { } && loweredThird.Type.IsStringType());
-            Debug.Assert(loweredFourth.HasAnyErrors || loweredFourth.Type is { } && loweredFourth.Type.IsStringType());
+            var firstType = loweredFirst.Type;
+            var secondType = loweredSecond.Type;
+            var thirdType = loweredThird.Type;
+            var fourthType = loweredFourth.Type;
+
+            Debug.Assert(loweredFirst.HasAnyErrors || firstType is not null && (firstType.IsStringType() || firstType.IsCharType() || firstType.IsReadOnlySpanChar()));
+            Debug.Assert(loweredSecond.HasAnyErrors || secondType is not null && (secondType.IsStringType() || secondType.IsCharType() || secondType.IsReadOnlySpanChar()));
+            Debug.Assert(loweredThird.HasAnyErrors || thirdType is not null && (thirdType.IsStringType() || thirdType.IsCharType() || thirdType.IsReadOnlySpanChar()));
+            Debug.Assert(loweredFourth.HasAnyErrors || fourthType is not null && (fourthType.IsStringType() || fourthType.IsCharType() || fourthType.IsReadOnlySpanChar()));
+
+            var firstIsChar = firstType?.IsCharType() == true;
+            var secondIsChar = secondType?.IsCharType() == true;
+            var thirdIsChar = thirdType?.IsCharType() == true;
+            var fourthIsChar = fourthType?.IsCharType() == true;
+
+            if ((firstIsChar || secondIsChar || thirdIsChar || fourthIsChar || firstType.IsReadOnlySpanChar() || secondType.IsReadOnlySpanChar() || thirdType.IsReadOnlySpanChar() || fourthType.IsReadOnlySpanChar()) &&
```

Comment:
>firstIsChar || secondIsChar || thirdIsChar || fourthIsChar || firstType.IsReadOnlySpanChar() || secondType.IsReadOnlySpanChar() || thirdType.IsReadOnlySpanChar() || fourthType.IsReadOnlySpanChar() [](http://example.com/codeflow?start=17&length=196)

Similar to the previous helper, could you point me to unit-tests that cover char and span at each specific position?